<h1><center>PORTADA</h1>

<h4>
<center>
    <h4>Ejercicio 1: Estructuras de Datos y DataFrames.</h4>
<h4>
Curso: Programación para el Analisis de datos<br>
Tarea: Principios básicos de programación científica<br>
Presentado por: Wagner Alonso Fernández Villalobosz<br>
Grupo: 203008069_14<br>
Código: 203008069A_1702<br>
Presentado a: Jose Laureano Cruz Cardozo<br>
Universidad: Universidad Nacional Abierta y a Distancia – UNAD<br>
Fecha: 24/04/2024<br>
<h/4>
</center>

<h1><center>DESARROLLO</h1>

<p style="text-align: justify;">Cargue en la variable df1 el conjunto de datos que se presentan en el archivo Excel, Anexo 1 - Base de Datos Excel.xlsx, que se encuentra en la carpeta de la Unidad 1 - Tarea 1 - Fundamentos de programación científica con Python. Tenga en cuenta la hoja, encabezado y celdas en las que comienzan los datos. Los índices del dataframe deben ser los valores en primera columna de la tabla de datos. </p>

1- Antes de leer los archivos, se procede con importar las librerías para leer los datos, basado en el material de Roldán et al. (2019):

In [1]:
import pandas as pd # pandas para lectura, manipulación de datos
from sklearn.datasets import load_breast_cancer# librería de datasets ejemplo del repositorio de Sckit-learn con datos sobre el cancer

Se carga la lectura del fichero en la variable df1 tal y como se encuentra el fichero descargado en el campus Vargas(2022):

In [2]:
#Lectura del archivo .csv en la variable df1, sin verificar
df1=pd.read_excel('data/Anexo 1 - Base de Datos EXCEL.xlsx ', sheet_name='BASE DE DATOS')#lectura del archivo Excel, seleccionando la hoja con datos
df1# instrucción a la computadora para la lectura del dataframe

Unnamed: 0 Unnamed: 1 Unnamed: 2        Unnamed: 3  \
0     BASE DE DATOS        NaN        NaN               NaN   
1               NaN        NaN        NaN               NaN   
2          Servicio    Empresa      Turno  Día de la semana   
3                 1        TAC          D                 5   
4                 2        TAD          N                 4   
...             ...        ...        ...               ...   
2198           2196        TAB          D                 3   
2199           2197        TAB          N                 6   
2200           2198        TAB          D                 3   
2201           2199        TAD          D                 7   
2202           2200        TAD          D                 3   

                     Unnamed: 4                Unnamed: 5  \
0                           NaN                       NaN   
1                           NaN                       NaN   
2     Experiencia del conductor  Formación del Paramédico   
3                  Principiante                         4   
4                  Principiante                         3   
...                         ...                       ...   
2198                    Experto                         3   
2199                 Intermedio                         4   
2200               Principiante                         2   
2201                 Intermedio                         1   
2202                 Intermedio                         1   

                Unnamed: 6              Unnamed: 7                Unnamed: 8  \
0                      NaN                     NaN                       NaN   
1                      NaN                     NaN                       NaN   
2     Nivel del Paramédico  Cantidad de atenciones  Distancia recorrida (Km)   
3                        I                       8                      78.4   
4                        B                       5                      46.2   
...                    ...                     ...                       ...   
2198                     B                       6                        58   
2199                     B                       7                      69.6   
2200                     B                       5                        50   
2201                     B                       6                      57.2   
2202                     I                       8                      84.1   

                   Unnamed: 9                       Unnamed: 10  \
0                         NaN                               NaN   
1                         NaN                               NaN   
2     Velocidad máxima (Km/h)  Consumo de combustible (Galones)   
3                          82                               3.1   
4                          89                               1.9   
...                       ...                               ...   
2198                      100                               1.7   
2199                       92                                 3   
2200                       86                                 2   
2201                       94                               2.6   
2202                       92                               3.6   

                   Unnamed: 11          Unnamed: 12  
0                          NaN                  NaN  
1                          NaN                  NaN  
2     Tiempo en Movimiento (h)  Tiempo motor ON (h)  
3                          2.8                 3.08  
4                          2.1                 2.31  
...                        ...                  ...  
2198                       2.9                 3.19  
2199                       2.9                 4.64  
2200                       2.5                 3.75  
2201                       2.6                 2.86  
2202                       2.9                 4.35  

[2203 rows x 13 columns]

Se realiza el ajuste para que el Dataframe se pueda visualizar adecuadamente, utilizando el metodo de la librería pandas llamado skiprows Álvarez (2020):

In [3]:
"""
Lectura del archivo .xlsx que se almacenada en la variable df1,
al leer el archivo Excel, se observa que las primeras 3 filas
del archivo ofimático son celdas en blanco; se procede con ajustar
con el método skiprows de Pandas para omitir y asignar nombres de columnas
mediante una lista.

"""

df1=pd.read_excel('data/Anexo 1 - Base de Datos EXCEL.xlsx', sheet_name='BASE DE DATOS', skiprows=3)#se aplica el metodo skiprows para omitir las 3 primeras filas

asignar_nombres_columnas = ['Servicio', 'Empresa', 'Turno', 'Día de la semana', 'Experiencia del conductor',
                    'Formación del Paramédico', 'Nivel del Paramédico', 'Cantidad de atenciones',
                    'Distancia recorrida (Km)', 'Velocidad máxima (Km/h)', 'Consumo de combustible (Galones)',
                    'Tiempo en Movimiento (h)', 'Tiempo motor ON (h)'] #se crea una lista para asignar nombres a las columnas del dataframe df1

df1.columns = asignar_nombres_columnas # se utiliza el metodo columns de pandas para asignar nombre a las columnas del df1

df1# instrucción a la computadora para la lectura del dataframe ya ajustado


Servicio Empresa Turno  Día de la semana Experiencia del conductor  \
0            1     TAC     D                 5              Principiante   
1            2     TAD     N                 4              Principiante   
2            3     TAB     N                 4                   Experto   
3            4     TAB     D                 2                Intermedio   
4            5     TAC     N                 7              Principiante   
...        ...     ...   ...               ...                       ...   
2195      2196     TAB     D                 3                   Experto   
2196      2197     TAB     N                 6                Intermedio   
2197      2198     TAB     D                 3              Principiante   
2198      2199     TAD     D                 7                Intermedio   
2199      2200     TAD     D                 3                Intermedio   

      Formación del Paramédico Nivel del Paramédico  Cantidad de atenciones  \
0                            4                    I                       8   
1                            3                    B                       5   
2                            4                    B                       6   
3                            3                    B                       7   
4                            2                    I                       8   
...                        ...                  ...                     ...   
2195                         3                    B                       6   
2196                         4                    B                       7   
2197                         2                    B                       5   
2198                         1                    B                       6   
2199                         1                    I                       8   

      Distancia recorrida (Km)  Velocidad máxima (Km/h)  \
0                         78.4                       82   
1                         46.2                       89   
2                         62.4                       99   
3                         70.2                       95   
4                         75.6                       87   
...                        ...                      ...   
2195                      58.0                      100   
2196                      69.6                       92   
2197                      50.0                       86   
2198                      57.2                       94   
2199                      84.1                       92   

      Consumo de combustible (Galones)  Tiempo en Movimiento (h)  \
0                                  3.1                       2.8   
1                                  1.9                       2.1   
2                                  2.5                       2.6   
3                                  3.0                       2.6   
4                                  2.8                       2.8   
...                                ...                       ...   
2195                               1.7                       2.9   
2196                               3.0                       2.9   
2197                               2.0                       2.5   
2198                               2.6                       2.6   
2199                               3.6                       2.9   

      Tiempo motor ON (h)  
0                    3.08  
1                    2.31  
2                    4.68  
3                    3.12  
4                    3.64  
...                   ...  
2195                 3.19  
2196                 4.64  
2197                 3.75  
2198                 2.86  
2199                 4.35  

[2200 rows x 13 columns]

<p style="text-align: justify;">  2-Cargue en la variable df2 el conjunto de datos que se presentan en el archivo CSV, Anexo 2 - Base de Datos CSV.csv, que se encuentra en la carpeta de la Unidad 1 - Tarea 1 - Fundamentos de programación científica con Python. 

Tenga en cuenta el tipo de separador y los encabezados de las columnas.</p>

Se procede con hacer la lectura del archivo .csv aplicando el método read_csv() y se agrega el delimitador que este caso es una coma (,) conforme a lo consultado en el material de Vargas (2022):

In [4]:
df2=pd.read_csv('data/Anexo 2 - Base de Datos CSV.csv',delimiter=',')#Lectura del archivo .csv que será cargado en la variable df2
df2# instrucción a la computadora para la lectura del dataframe 

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0       -122.05     37.37                27.0       3885.0           661.0   
1       -118.30     34.26                43.0       1510.0           310.0   
2       -117.81     33.78                27.0       3589.0           507.0   
3       -118.36     33.82                28.0         67.0            15.0   
4       -119.67       NaN                19.0       1241.0           244.0   
...         ...       ...                 ...          ...             ...   
2995    -119.86       NaN                23.0       1450.0           642.0   
2996    -118.14     34.06                27.0       5257.0          1082.0   
2997    -119.70     36.30                10.0        956.0           201.0   
2998    -117.12     34.10                40.0         96.0            14.0   
2999    -119.63     34.42                42.0       1765.0           263.0   

      population  households  median_income  median_house_value  
0         1537.0       606.0         6.6085            344700.0  
1          809.0       277.0         3.5990            176500.0  
2         1484.0       495.0         5.7934            270500.0  
3           49.0        11.0         6.1359            330000.0  
4          850.0       237.0         2.9375             81700.0  
...          ...         ...            ...                 ...  
2995      1258.0       607.0         1.1790            225000.0  
2996      3496.0      1036.0         3.3906            237200.0  
2997       693.0       220.0         2.2895             62000.0  
2998        46.0        14.0         3.2708            162500.0  
2999       753.0       260.0         8.5608            500001.0  

[3000 rows x 9 columns]

<p style="text-align: justify;"> 3-Cargue los datos Breast Cancer Wisconsin Dataset que se encuentran en la variable load_breast_cancer del paquete sklearn, en un dataframe de pandas asignado a la variable df3.</p>

In [5]:
#cargar el conjunto de datos Breast Cancer Wisconsin Dataset:
data = load_breast_cancer()# cargar el dataset con información sobre el canccer
df3 = pd.DataFrame(data.data, columns=data.feature_names)# se utiliza la función pd de pandas para convertirlo en DataFrame.
df3# instrucción a la computadora para la lectura del dataframe 

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

<p style="text-align: justify;">4-Construya un Panel con los tres (3) dataframes cargados anteriormente: df1, df2 y df3. Para construir este panel, cree un diccionario llamado panel que en sus keys tenga los nombres de los conjuntos de datos y los valores sean los dataframes.</p>

In [6]:
#construir un Panel con los tres DataFrames
panel = {'df1': df1, 'df2': df2, 'df3': df3}# creación del diccionario con sus llaves, donde están los nombres de los 
#anteriores dataframes cargados.

<p style="text-align: justify;">5-Programe una función que extraiga todas las listas de los encabezados de columnas que se encuentran en un Panel yúsela para extraer las listas de columnas del Panel del punto anterior. Imprima los resultados de forma que se diferencien los encabezados entre los dataframes en el Panel.</p>

Se procede con crear las respectivas funciones python para extraer las listas de los encabezados, según analizado en el material de Grus (2023):

In [7]:
"""
En el siguiente codigo, se define una función llamada extraer_encabezados de los 3
dataframes, cargados en las variables df1 df2 df3
"""

def extraer_encabezados(panel):# se define la función extraer_encabezados de todos los dataframes
    encabezados = {}
    for nombre, df in panel.items():# ciclo for que va a tomar todos titulos de las columnas de los dataframes 
        encabezados[nombre] = df.columns.tolist()# impresión de los datos en una lista
    return encabezados #devolver valores especificos de donde se llamó la función.


encabezados = extraer_encabezados(panel)# se crea una variable llamada encabezados, que invoque la función extraer_encabezados
print("Encabezados de las columnas en el Panel:")# impresión de mensaje
for nombre, cols in encabezados.items():# ciclo for para que proceda a imprimir los titulos de los 3 dataframes.
    print(f"{nombre}: {cols}") # impresión de mensaje

Encabezados de las columnas en el Panel:
df1: ['Servicio', 'Empresa', 'Turno', 'Día de la semana', 'Experiencia del conductor', 'Formación del Paramédico', 'Nivel del Paramédico', 'Cantidad de atenciones', 'Distancia recorrida (Km)', 'Velocidad máxima (Km/h)', 'Consumo de combustible (Galones)', 'Tiempo en Movimiento (h)', 'Tiempo motor ON (h)']
df2: ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'median_house_value']
df3: ['mean radius', 'mean texture', 'mean perimeter', 'mean area', 'mean smoothness', 'mean compactness', 'mean concavity', 'mean concave points', 'mean symmetry', 'mean fractal dimension', 'radius error', 'texture error', 'perimeter error', 'area error', 'smoothness error', 'compactness error', 'concavity error', 'concave points error', 'symmetry error', 'fractal dimension error', 'worst radius', 'worst texture', 'worst perimeter', 'worst area', 'worst smoothness', 'worst compactness', 'worst

<p style="text-align: justify;">6-Programe una función que imprima la cantidad de elementos que tienen las bases de datos en un Panel y úsela con el Panel del punto anterior. Imprima los resultados de forma que se diferencien los tamaños de los diferentes dataframes en el Panel.</p>

Se realliza el procedimiento conforme a lo consultado en el matierial de Grus (2023), que es crear funciones para imprimir las dimensiones de los DataFrame.

In [8]:
# Paso 5: Programar una función para imprimir la cantidad de elementos en cada DataFrame del Panel
def imprimir_tamanos(panel):# se define función para imprimit los tamaños o dimensiones del dataframe
    tamanos = {}
    for nombre, df in panel.items():# ciclo for para que tome los valores del panel y después la función shape para obtener los numeros de filas y columnas de los df
        tamanos[nombre] = df.shape
    return tamanos#devolver valores especificos de donde se llamó la función.

tamanos = imprimir_tamanos(panel)
print("Cantidad de elementos en cada DataFrame del Panel:")# impresión de mensaje
for nombre, tamano in tamanos.items():# ciclo for para que proceda a imprimir los tamaños de los conjuntos de datos de todo el panel
    print(f"{nombre}: {tamano[0]} filas, {tamano[1]} columnas")


Cantidad de elementos en cada DataFrame del Panel:
df1: 2200 filas, 13 columnas
df2: 3000 filas, 9 columnas
df3: 569 filas, 30 columnas


7-Filtre el dataframe df1 para obtener las muestras de la empresa TAC. Determine cuantos elementos tiene dicho filtro.

In [9]:
df1_filtrado = df1[df1['Empresa'] == 'TAC']# se utiliza una xxx tomar del dataframe la columna "Empresa" que filtre la palabra TAC

print("El DataFrame filtrado, en la columna 'EMPRESA' tiene muestras TAC:")# impresión de mensaje
print(f"Número de filas: {df1_filtrado.shape[0]}")# impresión de mensaje donde muestra la cantidad de filas en el df
print(f"Número de columnas: {df1_filtrado.shape[1]}")# impresión de mensaje donde muestra la cantidad de columnas en el df

El DataFrame filtrado, en la columna 'EMPRESA' tiene muestras TAC:
Número de filas: 596
Número de columnas: 13


8-Determine cuantos elementos nulos hay en cada una de las columnas del dataframe df2.

In [10]:
#Determinar cuántos elementos nulos hay en cada columna de df2
valores_nulos_por_columna = df2.isnull().sum()# se define variable donde se convoca una función llamada isnull() y .sum() para
#que detecte y sume los valores nulos en el df2.

print("Número de elementos nulos por columna en el df2:")# impresión de mensaje
print(valores_nulos_por_columna)# impresión de mensaje

suma_total_valores_nulos =valores_nulos_por_columna.sum()# sumtaoria de los valores


print('La sumatoria de los elementos en el DataFrame 2 (df2) es de: ',suma_total_valores_nulos, 'valores nulos en todas las filas y columnas')


Número de elementos nulos por columna en el df2:
longitude             0
latitude              5
housing_median_age    1
total_rooms           1
total_bedrooms        1
population            3
households            2
median_income         2
median_house_value    0
dtype: int64
La sumatoria de los elementos en el DataFrame 2 (df2) es de:  15 valores nulos en todas las filas y columnas


 <h1><center>ANALISIS DE LOS RESULTADOS</h1>

* Verificar las librerías y software necesario para desarrollar la actividad.

* Antes de manipular cualquier tipo de archivo, verificar si se visualiza adecuadamente y hacer los respectivos ajustes para que se vea organizado, de esto depende para hacer mejores visualizaciones graficas y que los algoritmos de aprendizaje automático hagan predicciones mejores.

* Los conjuntos de datos, tienen muchas dimensiones en columnas y filas que son: df1: 2200 filas, 13 columnas,df2: 3000 filas, 9 columnas y df3: 569 filas, 30 columnas.

* Las muestras TAC de la columna "EMPRESA" tiene un total de 15 valores nulos.

<h1><center>CONCLUSION</h1>

<p style="text-align: justify;">Con el apoyo del material de Roldán(2019) y lo consultado en el campus virtual, manipular los datos con Pandas en Python, es fundamental para cualquier profesional que ejerza roles data, la capacidad de cargar, limpiar, transformar y analizar datos de manera eficiente con la librería en mención, no solo aumenta la productividad, sino que también proporciona una bases sólidas para entender los datos y se puedan tomar decisiones informadas; conociendo las funcionalidades de Pandas, los desarrolladores especializados en datos, pueden explorar y entender rápidamente conjuntos de datos complejos, extrayendo información valiosa que impulsa el descubrimiento de conocimientos y la innovación en una amplia variedad de campos y disciplinas.</p>

<h1><center>BIBLIOGRAFÍA</h1>

Álvarez, C. (2020). Introducción al Jupyter Notebook y aplicaciones básicas. https://repository.unad.edu.co/handle/10596/35681

Grus, J. (2023). Ciencia de datos desde cero. Segunda edición: Principios básicos con Python. ANAYA MULTIMEDIA

Roldán, R. C., Martín, E. M., & Rodríguez, A. R. (2019). Big data con python: Recolección, almacenamiento y proceso. Alpha Editorial.

Vargas, M. (2022). OVI. Carga de Datos en Python desde Excel. [Objeto_virtual_de_Informacion_OVI].  
&nbsp;&nbsp;&nbsp;&nbsp;Repositorio Institucional &UNAD. https://repository.unad.edu.co/handle/10596/49755


